### PR TITLE
Update helm broker catalog on configmap change

### PR DESCRIPTION
**Pull request by polskikiel**
**Description**

Changes proposed in this pull request:

- HelmBroker will now watch at the configmaps labelled with `helm-broker-repo=true`and fetch bundles from the URLs defined in these configmaps, it's done by controller.
- Created pre-upgrade job to enter this changes in case of upgrade.

**Related issue(s)**

### CODE REVIEW

File: components/helm-broker/cmd/broker/main.go [hunk 1]
```diff
@@ -24,6 +23,16 @@ import (
...
 	"github.com/kyma-project/kyma/components/helm-broker/internal/storage"
 	"github.com/kyma-project/kyma/components/helm-broker/platform/logger"
+	v12 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	v1 "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	mapLabelKey       = "repo"
```
[the remark this answers is not in the document]
Comment (polskikiel) [reply components/helm-broker/cmd/broker/main.go]: will use "helm-broker-repo"

File: components/helm-broker/cmd/broker/main.go [hunk 2]
```diff
@@ -24,6 +23,16 @@ import (
 	"github.com/kyma-project/kyma/components/helm-broker/internal/helm"
 	"github.com/kyma-project/kyma/components/helm-broker/internal/storage"
 	"github.com/kyma-project/kyma/components/helm-broker/platform/logger"
+	v12 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
[the remark this answers is not in the document]
Comment (polskikiel) [reply components/helm-broker/cmd/broker/main.go]: done

File: components/helm-broker/cmd/broker/main.go [hunk 3]
```diff
@@ -53,20 +60,27 @@ func main() {
 	fatalOnError(err)
 	csbInterface := scClientSet.ServicecatalogV1beta1().ClusterServiceBrokers()
 
+	// broker sync
+	stopCh := make(chan struct{})
+	bLoader := bundle.NewLoader(cfg.TmpDir, log)
+	bundleSyncer := bundle.NewSyncer(sFact.Bundle(), sFact.Chart(), log)
 	brokerSyncer := broker.NewClusterServiceBrokerSyncer(csbInterface, log)
+	cfgMapInformer := v1.NewFilteredConfigMapInformer(clientset, cfg.Namespace, time.Minute, cache.Indexers{}, func(options *v12.ListOptions) {
```
[the remark this answers is not in the document]
Comment (polskikiel) [reply components/helm-broker/cmd/broker/main.go]: changed it to 10mins

File: components/helm-broker/cmd/preupgrade/main.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+	log := logger.New(&cfg.Logger)
+
+	log.Info("Searching for additional repositories")
+	deployments, err := clientset.AppsV1().Deployments(cfg.Namespace).List(v1.ListOptions{
+		LabelSelector: "app=helm-broker",
+	})
+	fatalOnError(err)
+
+	reposURLs := ""
+	for _, deploy := range deployments.Items {
```
[the remark this answers is not in the document]
Comment (polskikiel) [reply components/helm-broker/cmd/preupgrade/main.go]: Added the warning if such a situation occur

File: components/helm-broker/cmd/preupgrade/main.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+func newRestClientConfig(kubeConfigPath string) (*rest.Config, error) {
+	if kubeConfigPath != "" {
+		return clientcmd.BuildConfigFromFlags("", kubeConfigPath)
+	}
+
+	return rest.InClusterConfig()
+}
+
+func fixBundlesRepos(urls string) *corev1.ConfigMap {
```
[the remark this answers is not in the document]
Comment (polskikiel) [reply components/helm-broker/cmd/preupgrade/main.go]: changed to `migratePreviousBundlesRepos`

File: components/helm-broker/cmd/preupgrade/main.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+	return &corev1.ConfigMap{
+		TypeMeta: v1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "ConfigMap",
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name: "migrated-bundles-repos",
```
[the remark this answers is not in the document]
Comment (polskikiel) [reply components/helm-broker/cmd/preupgrade/main.go]: done

File: components/helm-broker/internal/broker/catalog_syncer.go [hunk 7]
```diff
@@ -6,28 +6,17 @@ import (
...
 		underlying: underlying,
 	}
 }
 
 // GetCatalog triggers syncer and execute underlying catalogService
 func (cs *catalogSyncerService) GetCatalog(ctx context.Context, osbCtx OsbContext) (*osb.CatalogResponse, error) {
-	// Trigger sync with bundle repositories
-	// If the sync is too long, the Service Catalog call timeout is exceeded and the Service Catalog
-	// will try again
-	cs.syncer.Execute()
```
[the remark this answers is not in the document]
Comment (polskikiel) [reply components/helm-broker/internal/broker/catalog_syncer.go]: reverted

File: components/helm-broker/cmd/broker/main.go [hunk 8]
```diff
@@ -24,6 +23,16 @@ import (
...
+	v12 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	v1 "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	mapLabelKey       = "repo"
+	mapLabelValue     = "true"
+	defaultMaxRetries = 5
```
[the remark this answers is not in the document]
Comment (polskikiel) [reply components/helm-broker/cmd/broker/main.go]: deleted

File: components/helm-broker/cmd/preupgrade/main.go [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+	for _, deploy := range deployments.Items {
+		for _, container := range deploy.Spec.Template.Spec.Containers {
+			for _, env := range container.Env {
+				if env.Name == "APP_REPOSITORY_URLS" {
+					reposURLs = env.Value
+				}
+			}
+		}
+	}
+	newURLs := strings.Join(strings.Split(reposURLs, ";"), "\n")
```
Comment (piotrmiskiewicz) [inline components/helm-broker/cmd/preupgrade/main.go]: filter URLs - default repo url must be removed from the list

Comment (polskikiel) [reply components/helm-broker/cmd/preupgrade/main.go]: fixed

**Pull request merged**